opencdms_process/process/climatol/__init__.py

```diff
@@ -19,29 +19,131 @@ def windrose(obs):
+        spclasses <- paste(vvc[1:nvvc],vvc[2:(nvvc+1)],sep='-')
+        geflag <- FALSE
+      }}
 
-    library("magick")
+      fr <- table(vvf,dd)
+
+      z <- as.integer(colnames(fr))
+      if(length(z)<16) {{
+        zrn <- rownames(fr)
+        zm <- matrix(rep(0,16*nrow(fr)),dim(fr))
+        zm[,z] <- fr
+        fr <- zm
+        colnames(fr) <- as.character(1:16)
+        rownames(fr) <- zrn
+      }}
+
+      nd1 <- sum(fr[1,]); fr[1,] <- fr[1,] * (nd1+cal)/nd1
+      if(nd>0) fr <- fr*100./nd
+      frt <- apply(fr,1,sum)
+      frtd <- apply(fr,2,sum)
+      tab <- cbind(fr,frt)
+      tab <- round(rbind(tab,c(frtd,sum(frtd))),1)
+      tab <- data.frame(rbind(tab,round(c(vm,vmt),1),round(c(vx,vmx),1)))
+      names(tab) <- c('N','NNE','NE','ENE','E','ESE','SE','SSE','S','SSW','SW',
+        'WSW','W','WNW','NW','NNW','Total')
+      row.names(tab) <- c(spclasses,'Total','Mean Sp.','Mx.M.Sp.')
+      old.par <- par(no.readonly = TRUE)
+      on.exit(par(old.par))
+      fr <- tab[1:nvvc,1:16]
+      if(geflag) row.names(fr)[nvvc] <- paste('>=',vvc[nvvc])
+      ndir <- 16
+      nr <- nvvc
+      fmax <- fnum*fint
+      key <- (nr>1)
+
+      if(key) mlf <- 3 else mlf <- 1
+      par(mar=margin, new=FALSE)
+      fx <- cos(pi/2-(2*pi/ndir*0:(ndir-1)))
+      fy <- sin(pi/2-(2*pi/ndir*0:(ndir-1)))
+      plot(fx,fy,xlim=c(-fmax-mlf*fint,fmax+fint),ylim=c(-fmax-fint,fmax+fint),
+        xaxt="n",yaxt="n",xlab="",ylab="",bty="n",asp=1,type="n")
+      if(nr==1) {{
+        cx <- fx*fr
+        cy <- fy*fr
+      }}
+      else {{
+        f <- apply(fr,2,sum)
+        cx <- fx*f
+        cy <- fy*f
+        for(i in nr:2) {{
+          f <- f-fr[i,]
+          cx <- c(cx,NA,fx*f)
+          cy <- c(cy,NA,fy*f)
+        }}
+      }}
+      polygon(cx,cy,col=col[nr:1])
+      symbols(c(0*1:fnum),c(0*1:fnum),circles=c(fint*1:fnum),inches=FALSE,add=TRUE)
+      segments(0*1:ndir,0*1:ndir,fmax*fx,fmax*fy)
+      fmaxi <- fmax+fint/4
+      text(0,fmaxi,"N")
+      text(0,-fmaxi,"S")
+      text(fmaxi,0,"E")
+      text(-fmaxi,0,"W")
+      if(flab==2)
+        for(i in 1:fnum) text(i*fint*cos(ang),i*fint*sin(ang),paste(i*fint,"%"))
+      else if(flab==1)
+        text(fmax*cos(ang),fmax*sin(ang),paste(fmax,"%"))
+      if(key) {{
+        legend(-fmaxi-2.3*fint,fmaxi+2,fill=col,legend=spclasses)
+        text(-fmaxi-1.5*fint,fmaxi+.9*fint,uni)
+      }}
+      title(paste(name,'windrose',startdate,'to',enddate))
+      invisible(tab)
+    }}
     figure <- image_graph(width = 350, height = 350, res = 96)
 
     data=observations
     ob_time=as.POSIXct(data$ob_time,tz='UTC')
     data=cbind(ob_time, data[,3:4])
-
-    windrose(data,"838","Bracknell Beaufort Park")
+    windrose(data,'838','Bracknell Beaufort Park')
 
     image <- image_write(figure, path = NULL, format = "png")
 
-    ''')
+    """)
     image_data = ro.globalenv['image']
     image = Image.open(BytesIO(bytes(image_data)))
-    return image
```

Comment:
`pyopencdms` and `opencdms_process` will be used by scientific programmers to access and analyse CDMS data in Python.

It looks like the original example code in the `pyopencdms-demo` repository, which return a windrose plot as an image in Python, will no longer work (instead - only the base64 encoded version can be returned)?

I'd expect the default to by the Python image for users who are just using the Python library.

If we have multiple processes in the future, perhaps each should inherit from an abstract base class `Process` and that base class could have a method `as_base64` (although we would want processes that don't return images to automatically return `NotImplemented` without having to remember to override the base class?)

Reply:
Understood. I will make necessary changes.

Reply:
I think we're going to end up with **generic** classes like `class Process()` and `class ImageProcess(Process)` in *pyopencdms*, e.g.  `from opencdms.process import ImageProcess`.

`as_base64()` would then be a method on `ImageProcess`. We could also have a `save()` method to write an image to disk etc.

We may then have **specific** classes like `Windrose(opencdms.process.ImageProcess)` located in *opencdms_process* in a climatol module (`from opencdms_process.climatol import Windrose`).

Our web code would then look something like:
```
windrose = Windrose(params)
return windrose.as_base64()
```